NPE while compiling an LMF-capable SAM inside an anonymous JS class

```
Caused by: java.lang.NullPointerException
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.synthesizeSAMWrapper(GenJSCode.scala:5111)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genAnonFunction(GenJSCode.scala:5100)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1948)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2977)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1757)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStat(GenJSCode.scala:1709)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genBlock$8(GenJSCode.scala:2976)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2976)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genLabelApply(GenJSCode.scala:2436)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2244)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStat(GenJSCode.scala:1709)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genCaseBody$1(GenJSCode.scala:3033)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genTranslatedMatch$3(GenJSCode.scala:3037)
        at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:739)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:738)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genTranslatedMatch(GenJSCode.scala:3017)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.translateMatch$1(GenJSCode.scala:2945)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2952)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2977)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genLabelApply(GenJSCode.scala:2436)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2244)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStat(GenJSCode.scala:1709)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genCaseBody$1(GenJSCode.scala:3033)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genCaseBody$1(GenJSCode.scala:3027)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genCaseBody$1(GenJSCode.scala:3023)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genTranslatedMatch$3(GenJSCode.scala:3037)
        at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:739)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:738)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genTranslatedMatch(GenJSCode.scala:3017)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.translateMatch$1(GenJSCode.scala:2945)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2952)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBody$1(GenJSCode.scala:1683)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethodDef(GenJSCode.scala:1688)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethodWithCurrentLocalNameScope$8(GenJSCode.scala:1499)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethodWithCurrentLocalNameScope$1(GenJSCode.scala:1461)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethodWithCurrentLocalNameScope(GenJSCode.scala:1406)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethod$1(GenJSCode.scala:1371)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope(JSEncoding.scala:54)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope$(JSEncoding.scala:50)
        at org.scalajs.core.compiler.GenJSCode.withNewLocalNameScope(GenJSCode.scala:31)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethod(GenJSCode.scala:1371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.gen$1(GenJSCode.scala:380)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genClass$3(GenJSCode.scala:371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genClass$3$adapted(GenJSCode.scala:371)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.gen$1(GenJSCode.scala:371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genClass(GenJSCode.scala:386)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genApplyNew$6(GenJSCode.scala:2398)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope(JSEncoding.scala:54)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope$(JSEncoding.scala:50)
        at org.scalajs.core.compiler.GenJSCode.withNewLocalNameScope(GenJSCode.scala:31)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$nestedGenerateClass$1(GenJSCode.scala:181)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.nestedGenerateClass(GenJSCode.scala:181)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genApplyNew$5(GenJSCode.scala:2398)
        at scala.Option.getOrElse(Option.scala:121)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApplyNew(GenJSCode.scala:2395)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2240)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2977)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1894)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genActualArgs$5(GenJSCode.scala:4526)
        at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:739)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:738)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genActualArgs(GenJSCode.scala:4517)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genNormalApply(GenJSCode.scala:2545)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2256)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genNormalApply(GenJSCode.scala:2545)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2256)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genIsAsInstanceOf(GenJSCode.scala:2288)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApplyTypeApply(GenJSCode.scala:2276)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2231)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1757)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStat(GenJSCode.scala:1709)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genBlock$8(GenJSCode.scala:2976)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2976)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBody$1(GenJSCode.scala:1683)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethodDef(GenJSCode.scala:1688)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethodWithCurrentLocalNameScope$8(GenJSCode.scala:1499)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethodWithCurrentLocalNameScope$1(GenJSCode.scala:1461)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethodWithCurrentLocalNameScope(GenJSCode.scala:1406)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethod$1(GenJSCode.scala:1371)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope(JSEncoding.scala:54)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope$(JSEncoding.scala:50)
        at org.scalajs.core.compiler.GenJSCode.withNewLocalNameScope(GenJSCode.scala:31)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethod(GenJSCode.scala:1371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.gen$1(GenJSCode.scala:380)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genClass$3(GenJSCode.scala:371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genClass$3$adapted(GenJSCode.scala:371)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.gen$1(GenJSCode.scala:371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genClass(GenJSCode.scala:386)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$apply$5(GenJSCode.scala:302)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$apply$4(GenJSCode.scala:289)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.apply(GenJSCode.scala:276)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:436)
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:429)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1(Global.scala:400)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1$adapted(Global.scala:400)
        at scala.collection.Iterator.foreach(Iterator.scala:929)
        at scala.collection.Iterator.foreach$(Iterator.scala:929)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
        at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:400)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.run(GenJSCode.scala:217)
        at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1452)
        at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1436)
        at scala.tools.nsc.Global$Run.compileSources(Global.scala:1429)
        at scala.tools.nsc.Global$Run.compile(Global.scala:1545)
        at xsbt.CachedCompiler0.run(CompilerInterface.scala:116)
        at xsbt.CachedCompiler0.run(CompilerInterface.scala:95)
        at xsbt.CompilerInterface.run(CompilerInterface.scala:26)
        at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:101)
        at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:47)
        at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:41)
        at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply$mcV$sp(AggressiveCompile.scala:97)
        at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply(AggressiveCompile.scala:97)
        at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply(AggressiveCompile.scala:97)
        at sbt.compiler.AggressiveCompile.sbt$compiler$AggressiveCompile$$timed(AggressiveCompile.scala:162)
        at sbt.compiler.AggressiveCompile$$anonfun$3.compileScala$1(AggressiveCompile.scala:96)
        at sbt.compiler.AggressiveCompile$$anonfun$3.apply(AggressiveCompile.scala:139)
        at sbt.compiler.AggressiveCompile$$anonfun$3.apply(AggressiveCompile.scala:86)
        at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:38)
        at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:36)
        at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:31)
        at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:39)
        at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:38)
        at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:66)
        at sbt.inc.Incremental$.compile(Incremental.scala:38)
        at sbt.inc.IncrementalCompile$.apply(Compile.scala:26)
        at sbt.compiler.AggressiveCompile.compile2(AggressiveCompile.scala:153)
        at sbt.compiler.AggressiveCompile.compile1(AggressiveCompile.scala:70)
        at com.typesafe.zinc.Compiler.compile(Compiler.scala:201)
        at com.typesafe.zinc.Compiler.compile(Compiler.scala:183)
        at com.typesafe.zinc.Compiler.compile(Compiler.scala:174)
        at com.typesafe.zinc.Compiler.compile(Compiler.scala:165)
        at org.gradle.api.internal.tasks.scala.ZincScalaCompiler$Compiler.execute(ZincScalaCompiler.java:92)
        at org.gradle.api.internal.tasks.scala.ZincScalaCompiler.execute(ZincScalaCompiler.java:66)
        at org.gradle.api.internal.tasks.scala.ZincScalaCompiler.execute(ZincScalaCompiler.java:48)
        at org.gradle.api.internal.tasks.compile.daemon.CompilerDaemonServer.execute(CompilerDaemonServer.java:31)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:87)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:377)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)

```

--- a/compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala
+++ b/compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala
@@ -136,7 +136,7 @@ abstract class GenJSCode extends plugins.PluginComponent
 
     // Per method body
     private val currentMethodSym = new ScopedVar[Symbol]
-    private val thisLocalVarInfo = new ScopedVar[Option[(js.Ident, jstpe.Type)]]
+    private val thisLocalVarIdent = new ScopedVar[Option[js.Ident]]
     private val fakeTailJumpParamRepl = new ScopedVar[(Symbol, Symbol)]
     private val enclosingLabelDefParams = new ScopedVar[Map[Symbol, List[Symbol]]]
     private val isModuleInitialized = new ScopedVar[VarBox[Boolean]]
@@ -181,9 +181,9 @@ abstract class GenJSCode extends plugins.PluginComponent
       withScopedVars(
           currentClassSym := clsSym,
           unexpectedMutatedFields := mutable.Set.empty,
-          generatedSAMWrapperCount := null,
+          generatedSAMWrapperCount := new VarBox(0),
           currentMethodSym := null,
-          thisLocalVarInfo := null,
+          thisLocalVarIdent := null,
           fakeTailJumpParamRepl := null,
           enclosingLabelDefParams := null,
           isModuleInitialized := null,
@@ -1481,7 +1481,7 @@ abstract class GenJSCode extends plugins.PluginComponent
 
       withScopedVars(
           currentMethodSym          := sym,
-          thisLocalVarInfo          := None,
+          thisLocalVarIdent         := None,
           fakeTailJumpParamRepl     := (NoSymbol, NoSymbol),
           enclosingLabelDefParams   := Map.empty,
           isModuleInitialized       := new VarBox(false),
@@ -1731,7 +1731,16 @@ abstract class GenJSCode extends plugins.PluginComponent
 
           initialThis match {
             case Ident(_) =>
-              // TODO Is this special-case really needed?
+              /* This case happens in trait implementation classes, until
+               * Scala 2.11. In the method, all usages of `this` have been
+               * replaced by the method's formal parameter `$this`. However,
+               * there is still a declaration of the pseudo local variable
+               * `_$this`, which is used in the param list of the label. We
+               * need to remember it now, so that when we build the JS version
+               * of the formal params for the label, we can redirect the
+               * assignment to `$this` instead of the otherwise unused
+               * `_$this`.
+               */
               withScopedVars(
                 fakeTailJumpParamRepl := (thisDef.symbol, initialThis.symbol)
               ) {
@@ -1744,15 +1753,33 @@ abstract class GenJSCode extends plugins.PluginComponent
                 mutableLocalVars += thisSym
 
               val thisLocalIdent = encodeLocalSym(thisSym)
-              val thisLocalType = toIRType(thisSym.tpe)
+              val thisLocalType = currentClassType
 
-              val genRhs = genExpr(initialThis)
+              val genRhs = {
+                /* #3267 In default methods, scalac will type its _$this
+                 * pseudo-variable as the *self-type* of the enclosing class,
+                 * instead of the enclosing class type itself. However, it then
+                 * considers *usages* of _$this as if its type were the
+                 * enclosing class type. The latter makes sense, since it is
+                 * compiled as `this` in the bytecode, which necessarily needs
+                 * to be the enclosing class type. Only the declared type of
+                 * _$this is wrong.
+                 *
+                 * In our case, since we generate an actual local variable for
+                 * _$this, we must make sure to type it correctly, as the
+                 * enclosing class type. However, this means the rhs of the
+                 * ValDef does not match, which is why we have to adapt it
+                 * here.
+                 */
+                forceAdapt(genExpr(initialThis), thisLocalType)
+              }
+
               val thisLocalVarDef = js.VarDef(thisLocalIdent,
                   thisLocalType, thisSym.isMutable, genRhs)
 
               val innerBody = {
                 withScopedVars(
-                  thisLocalVarInfo := Some((thisLocalIdent, thisLocalType))
+                  thisLocalVarIdent := Some(thisLocalIdent)
                 ) {
                   genInnerBody()
                 }
@@ -1772,7 +1799,7 @@ abstract class GenJSCode extends plugins.PluginComponent
           if (currentClassSym.isImplClass) {
             val thisParam = jsParams.head
             withScopedVars(
-                thisLocalVarInfo := Some((thisParam.name, thisParam.ptpe))
+                thisLocalVarIdent := Some(thisParam.name)
             ) {
               genBody()
             }
@@ -1787,7 +1814,7 @@ abstract class GenJSCode extends plugins.PluginComponent
 
         val thisLocalIdent = freshLocalIdent("this")
         withScopedVars(
-          thisLocalVarInfo := Some((thisLocalIdent, jstpe.AnyType))
+          thisLocalVarIdent := Some(thisLocalIdent)
         ) {
           val thisParamDef = js.ParamDef(thisLocalIdent,
               jstpe.AnyType, mutable = false, rest = false)
@@ -1795,6 +1822,32 @@ abstract class GenJSCode extends plugins.PluginComponent
           js.MethodDef(static = true, methodName, thisParamDef :: jsParams,
               resultIRType, Some(genBody()))(
               optimizerHints, None)
+        }
+      }
+    }
+
+    /** Forces the given `tree` to a given type by adapting it.
+     *
+     *  @param tree
+     *    The tree to adapt.
+     *  @param tpe
+     *    The target type, which must be either `AnyType` or `ClassType(_)`.
+     */
+    private def forceAdapt(tree: js.Tree, tpe: jstpe.Type): js.Tree = {
+      if (tree.tpe == tpe || tpe == jstpe.AnyType) {
+        tree
+      } else {
+        /* Remove the useless cast that scalac's erasure had to introduce to
+         * work around their own ill-typed _$this. Note that the optimizer will
+         * not be able to do that, since it won't be able to prove that the
+         * underlying expression is indeed an instance of `tpe`.
+         */
+        tree match {
+          case js.AsInstanceOf(underlying, _) if underlying.tpe == tpe =>
+            underlying
+          case _ =>
+            val cls = tpe.asInstanceOf[jstpe.ClassType].className
+            js.AsInstanceOf(tree, jstpe.ClassRef(cls))(tree.pos)
         }
       }
     }
@@ -1871,7 +1924,10 @@ abstract class GenJSCode extends plugins.PluginComponent
         /** Local val or var declaration */
         case ValDef(_, name, _, rhs) =>
           /* Must have been eliminated by the tail call transform performed
-           * by genMethodDef(). */
+           * by genMethodDef().
+           * If you ever change/remove this assertion, you need to update
+           * genEnclosingLabelApply() regarding `nme.THIS`.
+           */
           assert(name != nme.THIS,
               s"ValDef(_, nme.THIS, _, _) found at ${tree.pos}")
 
@@ -2092,15 +2148,21 @@ abstract class GenJSCode extends plugins.PluginComponent
      *  is one.
      */
     private def genThis()(implicit pos: Position): js.Tree = {
-      thisLocalVarInfo.fold[js.Tree] {
+      thisLocalVarIdent.fold[js.Tree] {
         if (tryingToGenMethodAsJSFunction) {
           throw new CancelGenMethodAsJSFunction(
               "Trying to generate `this` inside the body")
         }
         js.This()(currentClassType)
-      } { case (thisLocalVarIdent, thisLocalVarTpe) =>
+      } { thisLocalIdent =>
         // .copy() to get the correct position
-        js.VarRef(thisLocalVarIdent.copy())(thisLocalVarTpe)
+        val tpe = {
+          if (currentClassSym.isImplClass)
+            encodeClassType(traitOfImplClass(currentClassSym))
+          else
+            currentClassType
+        }
+        js.VarRef(thisLocalIdent.copy())(tpe)
       }
     }
 
@@ -2605,23 +2667,49 @@ abstract class GenJSCode extends plugins.PluginComponent
 
       // Prepare quadruplets of (formalArg, irType, tempVar, actualArg)
       // Do not include trivial assignments (when actualArg == formalArg)
-      val formalArgs = enclosingLabelDefParams(sym)
-      val actualArgs = args map genExpr
       val quadruplets = {
-        for {
-          (formalArgSym, actualArg) <- formalArgs zip actualArgs
-          formalArg = encodeLocalSym(formalArgSym)
-          if (actualArg match {
-            case js.VarRef(`formalArg`) => false
-            case _                      => true
-          })
-        } yield {
-          mutatedLocalVars += formalArgSym
-          val tpe = toIRType(formalArgSym.tpe)
-          (js.VarRef(formalArg)(tpe), tpe,
-              freshLocalIdent("temp$" + formalArg.name),
-              actualArg)
+        val formalArgs = enclosingLabelDefParams(sym)
+        val quadruplets =
+          List.newBuilder[(js.VarRef, jstpe.Type, js.Ident, js.Tree)]
+
+        for ((formalArgSym, arg) <- formalArgs.zip(args)) {
+          val formalArg = encodeLocalSym(formalArgSym)
+          val actualArg = genExpr(arg)
+
+          /* #3267 The formal argument representing the special `this` of a
+           * tailrec method can have the wrong type in the scalac symbol table.
+           * We need to patch it up, along with the actual argument, to be the
+           * enclosing class type.
+           * See the longer comment in genMethodDef() for more details.
+           *
+           * Note that only testing the `name` against `nme.THIS` is safe,
+           * given that `genStatOrExpr()` for `ValDef` asserts that no local
+           * variable named `nme.THIS` can happen, other than the ones
+           * generated for tailrec methods.
+           */
+          val isTailJumpThisLocalVar = formalArgSym.name == nme.THIS
+
+          val tpe =
+            if (isTailJumpThisLocalVar) currentClassType
+            else toIRType(formalArgSym.tpe)
+
+          val fixedActualArg =
+            if (isTailJumpThisLocalVar) forceAdapt(actualArg, tpe)
+            else actualArg
+
+          actualArg match {
+            case js.VarRef(`formalArg`) =>
+              // This is trivial assignment, we don't need it
+
+            case _ =>
+              mutatedLocalVars += formalArgSym
+              quadruplets += ((js.VarRef(formalArg)(tpe), tpe,
+                  freshLocalIdent("temp$" + formalArg.name),
+                  fixedActualArg))
+          }
         }
+
+        quadruplets.result()
       }
 
       // The actual jump (continue labelDefIdent;)
@@ -5540,10 +5628,11 @@ abstract class GenJSCode extends plugins.PluginComponent
     sym.hasAnnotation(JSNativeAnnotation)
 
   /** Tests whether the given class is the impl class of a raw JS trait. */
-  private def isRawJSImplClass(sym: Symbol): Boolean = {
-    sym.isImplClass &&
-    isJSType(sym.owner.info.decl(sym.name.dropRight(nme.IMPL_CLASS_SUFFIX.length)))
-  }
+  private def isRawJSImplClass(sym: Symbol): Boolean =
+    sym.isImplClass && isJSType(traitOfImplClass(sym))
+
+  private def traitOfImplClass(sym: Symbol): Symbol =
+    sym.owner.info.decl(sym.name.dropRight(nme.IMPL_CLASS_SUFFIX.length))
 
   /** Tests whether the given member is exposed, i.e., whether it was
    *  originally a public or protected member of a non-native JS class.

--- a/project/BinaryIncompatibilities.scala
+++ b/project/BinaryIncompatibilities.scala
@@ -25,7 +25,10 @@ object BinaryIncompatibilities {
   val SbtPlugin = Seq(
   )
 
-  val TestAdapter = Seq(
+  val TestCommon = Seq(
+  )
+
+  val TestAdapter = TestCommon ++ Seq(
   )
 
   val CLI = Seq(
@@ -34,6 +37,6 @@ object BinaryIncompatibilities {
   val Library = Seq(
   )
 
-  val TestInterface = Seq(
+  val TestInterface = TestCommon ++ Seq(
   )
 }

--- a/project/Build.scala
+++ b/project/Build.scala
@@ -110,7 +110,7 @@ object Build {
     "Whether we should partest the current scala version (and fail if we can't)")
 
   /* MiMa configuration -- irrelevant while in 1.0.0-SNAPSHOT.
-  val previousVersion = "0.6.21"
+  val previousVersion = "0.6.22"
   val previousSJSBinaryVersion =
     ScalaJSCrossVersion.binaryScalaJSVersion(previousVersion)
   val previousBinaryCrossVersion =

--- a/test-suite/js/src/test/require-sam/org/scalajs/testsuite/jsinterop/SAMJSTest.scala
+++ b/test-suite/js/src/test/require-sam/org/scalajs/testsuite/jsinterop/SAMJSTest.scala
@@ -1,0 +1,43 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js Test Suite        **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2018, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package org.scalajs.testsuite.jsinterop
+
+import scala.scalajs.js
+
+import org.junit.Assert._
+import org.junit.Test
+
+import org.scalajs.testsuite.utils.JSAssert._
+
+class SAMJSTest {
+
+  import SAMJSTest._
+
+  @Test def samNestedInAnonJSClass_issue3264(): Unit = {
+    val outer = new SAMInAnonJSClass_ParentJSClass {
+      def foo(x: Int): Int = {
+        val innerSAM: SAMInAnonJSClass_MySAM = _ * 2
+        innerSAM.bar(x + 3)
+      }
+    }
+
+    assertEquals(16, outer.foo(5))
+  }
+
+}
+
+object SAMJSTest {
+  abstract class SAMInAnonJSClass_ParentJSClass extends js.Object {
+    def foo(x: Int): Int
+  }
+
+  trait SAMInAnonJSClass_MySAM {
+    def bar(x: Int): Int
+  }
+}

--- a/test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/RegressionTest.scala
+++ b/test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/RegressionTest.scala
@@ -691,6 +691,56 @@ class RegressionTest {
     assertEquals(15, new Child().bar(1, 0))
   }
 
+  @Test def tailrec_in_trait_with_self_type_scala_2_12_issue_3267(): Unit = {
+    class Parser {
+      def c(): Int = 65
+    }
+
+    trait Helpers { this: Parser =>
+      @tailrec
+      final def rec(i: Int): Int = {
+        if (i == 0) b() + c()
+        else rec(i - 1)
+      }
+
+      def b(): Int = 42
+    }
+
+    class ParserWithoutHelpers extends Parser {
+      def foo(): Int = 5
+    }
+
+    class ParserWithHelpers extends Parser with Helpers
+
+    assertEquals(5, new ParserWithoutHelpers().foo())
+    assertEquals(107, new ParserWithHelpers().rec(3))
+  }
+
+  @Test def tailrec_in_class_with_self_type_scala_2_12_issue_3267(): Unit = {
+    trait Parser {
+      def c(): Int = 65
+    }
+
+    class Helpers { this: Parser =>
+      @tailrec
+      final def rec(i: Int): Int = {
+        if (i == 0) b() + c()
+        else rec(i - 1)
+      }
+
+      def b(): Int = 42
+    }
+
+    class ParserWithoutHelpers extends Parser {
+      def foo(): Int = 5
+    }
+
+    class ParserWithHelpers extends Helpers with Parser
+
+    assertEquals(5, new ParserWithoutHelpers().foo())
+    assertEquals(107, new ParserWithHelpers().rec(3))
+  }
+
 }
 
 object RegressionTest {

--- a/test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnCheckedListTest.scala
+++ b/test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnCheckedListTest.scala
@@ -47,9 +47,9 @@ trait CollectionsCheckedListTest
   @Test def testCheckedList(): Unit = {
     superList().add(0, new C)
     assertTrue(superList().addAll(0, Seq(new C)))
-    testOnFirstPositionOfIterator[ju.ListIterator[A]](superList().listIterator,
+    testOnFirstPositionOfIterator[ju.ListIterator[A]](superList().listIterator _,
         _.add(new C), None)
-    testOnFirstPositionOfIterator[ju.ListIterator[A]](superList().listIterator,
+    testOnFirstPositionOfIterator[ju.ListIterator[A]](superList().listIterator _,
         _.set(new C), None)
   }
 
@@ -60,10 +60,10 @@ trait CollectionsCheckedListTest
     expectThrows(classOf[ClassCastException],
         superList().addAll(0, Seq(new A)))
     testOnFirstPositionOfIterator[ju.ListIterator[A]](
-        superList().listIterator,
+        superList().listIterator _,
         _.add(new A), Some(classOf[ClassCastException]))
     testOnFirstPositionOfIterator[ju.ListIterator[A]](
-        superList().listIterator,
+        superList().listIterator _,
         _.set(new A), Some(classOf[ClassCastException]))
   }
 

